extensions/typescript-language-features/src/languageFeatures/refactor.ts

```diff
@@ -516,16 +541,24 @@ class TypeScriptRefactorProvider implements vscode.CodeActionProvider<TsCodeActi
 			return undefined;
 		}
 
-		const actions = Array.from(this.convertApplicableRefactors(document, response.body, rangeOrSelection)).filter(action => {
+		const applicableRefactors = await this.convertApplicableRefactors(document, response.body, rangeOrSelection);
+		const actions = await Promise.all(applicableRefactors.map(async (action) => {
 			if (this.client.apiVersion.lt(API.v430)) {
 				// Don't show 'infer return type' refactoring unless it has been explicitly requested
 				// https://github.com/microsoft/TypeScript/issues/42993
 				if (!context.only && action.kind?.value === 'refactor.rewrite.function.returnType') {
 					return false;
 				}
 			}
+			if (action.kind?.value === Move_NewFile.kind.value) {
```

Comment:
Let's also scope these new checks to TS 5.3 and add a comment about removing them in the future

TS 5.4 is still in pre-release and that's where the fix on the TS side will go